web_monitoring/cli.py

```diff
@@ -1,51 +1,455 @@
+                                                           tags=self.tags,
+                                                           view_url=record.view_url)
+        except Exception as error:
+            # On connection failures, reset the session and try again. If we
+            # don't do this, the connection pool for this thread is pretty much
+            # dead. It's not clear to me whether there is a problem in urllib3
+            # or Wayback's servers that requires this.
+            # This unfortunately requires string checking because the error can
+            # get wrapped up into multiple kinds of higher-level errors :(
+            if retry_connection_failures and ('failed to establish a new connection' in str(error).lower()):
+                self.wayback.session.reset()
+                return self.process_record(record)
+
+            # Otherwise, re-raise the error.
+            raise error
+
+    @classmethod
+    def create_summary(cls):
+        """
+        Create a dictionary that summarizes the results of processing all the
+        CDX records on a queue.
+        """
+        return {'total': 0, 'success': 0, 'playback': 0, 'missing': 0,
+                'unknown': 0}
+
+    @classmethod
+    def summarize(cls, workers, initial=None):
+        """
+        Combine the summaries from multiple `WaybackRecordsWorker` instances
+        into a single summary.
+        """
+        return cls.merge_summaries((w.summary for w in workers), initial)
+
+    @classmethod
+    def merge_summaries(cls, summaries, intial=None):
+        merged = intial or cls.create_summary()
+        for summary in summaries:
+            for key in merged.keys():
+                if key in summary:
+                    merged[key] += summary[key]
+
+        # Add percentage calculations
+        if merged['total']:
+            merged.update({f'{k}_pct': 100 * v / merged['total']
+                           for k, v in merged.items()
+                           if k != 'total' and not k.endswith('_pct')})
+        else:
+            merged.update({f'{k}_pct': 0.0
+                           for k, v in merged.items()
+                           if k != 'total' and not k.endswith('_pct')})
+
+        return merged
+
+    @classmethod
+    def parallel(cls, count, *args, **kwargs):
+        """
+        Run several `WaybackRecordsWorker` instances in parallel. When this
+        returns, the workers will have finished running.
+
+        Parameters
+        ----------
+        count: int
+            Number of instances to run in parallel.
+        *args
+            Arguments to pass to each instance.
+        **kwargs
+            Keyword arguments to pass to each instance.
+
+        Returns
+        -------
+        list of WaybackRecordsWorker
+        """
+        workers = []
+        for i in range(count):
+            worker = cls(*args, **kwargs)
+            workers.append(worker)
+            worker.start()
+
```

Comment:
Random thought: Should this be a `contextlib.contextmanager` that yields `workers` and joins them on exit? If that's not useful in practice anyway, forget about it.

Reply:
For this use case, probably not. There’s nothing useful to be done with these workers while they’re running, but we need access to their summaries after they’re done. It seemed usefully flexible to return the completed workers instead of just their summaries in case we added other properties or something.